### PR TITLE
Feature/core overhaul

### PR DESCRIPTION
**Pull request by GlynLeine**

### CODE REVIEW

File: args/core/containers/atomic_sparse_map.hpp [hunk 1]
```diff
@@ -4,551 +4,665 @@
...
+
+        /**@brief Returns whether the sparse_map is empty.
+         * @returns bool True if the sparse_map is empty, otherwise false.
+         */
+        A_NODISCARD inline bool empty(std::memory_order order = std::memory_order_acquire) const noexcept { return m_size.load(order) == 0; }
+
+        /**@brief Clears sparse_map.
+         * @note Will not update capacity.
+         */
+        inline void clear(std::memory_order order = std::memory_order_release) noexcept { m_size.store(0, order); }
```
Comment (Algo-ryth-mix) [inline args/core/containers/atomic_sparse_map.hpp:127]: these new parameters need a `@param` field in the comments

**Review by Developer-The-Great — APPROVED** [review]

File: args/core/ecs/ecsregistry.hpp [hunk 2]
```diff
@@ -20,225 +20,311 @@ namespace args::core::ecs
...
+        A_NODISCARD component_container_base* getFamily(id_type componentTypeId);
+
+        /**@brief Get component handle of a certain type attached to a certain entity.
+         * @param entityId Id of entity to get the component from.
+         * @param componentTypeId Type id of component to get handle of.
+         * @returns component_handle_base Component handle to the requested component. (may be invalid if the entity does not have this component type)
+         * @note Handle needs to force_cast to component_handle<T> in order to use correctly.
+         * @throws args_entity_not_found_error If the entity id does not belong to a valid entity.
+         */
+        A_NODISCARD bool hasComponent(id_type entityId, id_type componentTypeId);
```
Comment (Algo-ryth-mix) [inline args/core/ecs/ecsregistry.hpp]: wrong comment

File: args/core/ecs/ecsregistry.hpp [hunk 3]
```diff
@@ -20,225 +20,311 @@ namespace args::core::ecs
...
             return force_value_cast<component_handle<component_type>>(createComponent(entityId, typeHash<component_type>()));
-		}
+        }
+
+        template<typename component_type, typename = doesnt_inherit_from<component_type, archetype_base>>
+        component_handle<std::remove_reference_t<component_type>> createComponent(id_type entityId, component_type&& component)
+        {
+            std::remove_reference_t<component_type> temp = component;
+            return force_value_cast<component_handle<std::remove_reference_t<component_type>>>(createComponent(entityId, typeHash<std::remove_reference_t<component_type>>(), &temp));
+        }
```
Comment (Algo-ryth-mix) [inline args/core/ecs/ecsregistry.hpp]: this line should probably be split

File: args/core/ecs/entity_handle.hpp [hunk 4]
```diff
@@ -217,6 +258,17 @@ namespace args::core::ecs
...
+         * @param componentTypeId Type id of component to add.
+         * @param value Pointer to component_type that has the starting value you require.
+         * @throws args_invalid_entity_error Thrown when handle's registry reference is invalid.
+         * @throws args_entity_not_found_error Thrown when handle's id is invalid.
+         * @throws args_unknown_component_error Thrown when the component type is unknown.
+         * @returns component_handle_base Valid component handle base for the newly created component.
+         * @note component_handle_base needs to be force_cast to component_handle<T> in order to be usable.
+         */
+        component_handle_base add_component(id_type componentTypeId, void* value) const;
+
```
Comment (Algo-ryth-mix) [inline args/core/ecs/entity_handle.hpp:271]: component_handle_base should either be a proxy or have a builtin converter to reduce the overhead of casting to the required type

**Review by Algo-ryth-mix — APPROVED** [review]

**Pull request merged**